lib/vsc/mympirun/mpi/mpi.py

```diff
@@ -572,8 +574,21 @@ def make_node_file(self):
             self.log.raiseException(msg)
 
     def get_universe_ncpus(self):
-        """Return ppn for universe"""
-        return self.mpiprocesspernode
+        """Return ppn dict for universe"""
+        universe_ppn = {}
+        for node in nub(self.nodes):
+            universe_ppn[node] = 0
+        i = 0
+        while i < self.options.universe:
+            for node, ppn in universe_ppn.items():
+                if ppn < self.ppn_dict[node]:
+                    universe_ppn[node] = ppn+1
+                    i += 1
+                if i >= self.options.universe:
+                    break
```

Comment:
not a big fan of basically having the same condition in two places (here + in `while`)...

maybe something like this?

but write a test first for your current implementation :)

```python
nodes = nub(self.nodes)
universe_ppn = dict((node, 0) for node in nodes)
proc_cnt = 0
node = nodes.pop(0)
while proc_cnt < self.options.universe:
    if universe_ppn[node] < self.ppn_dict[node]:
        universe_ppn[node] += 1
        proc_cnt += 1
        # recycle node
        nodes.append(node)
    # select next node to assign a process to
    node = nodes.pop(0)
```

You'll need to be careful with the last bit, in case `nodes` becomes empty (nodes are only recycled if a process was assigned).